### ISSUE TITLE
about the accuracy of training

### ISSUE THREAD

**Issue by bjlgcxc** [inspirehep/magpie#84]
how is the accuracy computed when the model is training(multi label text classification)?
why the accuracy of training is always greater than 90%?
Thank you for helping me.

**Comment by bjlgcxc** [comment]
and the accuracy of testing is very low

**Comment by NowayIndustries** [comment]
**Disclaimer**: I've only just started playing with magpie & deep learning and have not used it for anything big/serious yet (soon tough). Thus I can not offer any sort of guarantee that what I'm about to say will be right or even complete, perhaps @jstypka can validate my answer so we won't both be misguided by my attempt to answer you.

---

First off, please read[ this very good stack overflow answer](http://stackoverflow.com/a/34705152) and the answer of 'farizrahman4u' on[ this github issue](https://github.com/fchollet/keras/issues/877). This should give you some insight into interpreting the keras output.

So now we understand: a low accuracy of testing with a high accuracy of training is the result of the model you trained, it is unable to predict for text it has never seen before. This thus means that the model needs more training, because whilst it can work with text it knows (or has "memorized") I assume you want to predict labels for text you haven't trained on.

What training and testing data are you using? I saw the same results when testing magpie on the data that comes with this repo in the [/data/hep-keywords/data](https://github.com/inspirehep/magpie/tree/master/data/hep-keywords/data) directory.


---
_Just in case you want to have a peek at some source code:_
[In the MagPie model, in the train function](https://github.com/inspirehep/magpie/blob/master/magpie/__init__.py#L72) the keras function 'fit' is called with the train function's parameter 'test_dir' translated to test_data which is fed to keras' 'fit' as 'validation_data'.

[From the Keras documentation:](https://keras.io/models/model/#fit) **validation_data:** data on which to evaluate the loss and any model metrics at the end of each epoch. The model will not be trained on this data.

**Comment by bjlgcxc** [comment]
thank you for your answer

**Comment by bjlgcxc** [comment]
the data i use is from Mulan（multi-label classifaction framework） 
it contains many bookmarks（multi-label） and has more than 80000 instances

**Comment by NowayIndustries** [comment]
Interesting dataset(s), how are you converting the xml and arff to the format magpie expects?

From my quick peek at the bookmarks dataset it just looks like a whole bunch of numbers between curly braces and a xml scheme that defines tags.

**Comment by bjlgcxc** [comment]
an instance -> a file（or a book）, the file contains the words that its attribute value is 1,the same way for the labels。

first step,read the arff file by the Class（ArffLoader） that Weka provides,
second,use every instance that ArffLoader reads to rebuild every file and its label file（the .txt and .lab）
third,train the magpie model(the list of labels can be got from the .xml file)

**Comment by bjlgcxc** [comment]
public class DataConverter {
	
	private String arffFile;
	private String xmlFile;
	
	private List<String> attributes;
	private List<String> labels;
	
	private ArffLoader arffLoader;
	private Instances structure;
	private CSVSaver csvSaver;	
	
	/*
	 * init
	 */
	public DataConverter(String arffFile,String xmlFile) throws IOException, DocumentException {
		this.arffFile = arffFile;
		this.xmlFile = xmlFile;
		attributes = new ArrayList<String>();
		labels = new ArrayList<String>();
		setArffLoader();
		setCsvSaver();
		setLabels();
		setAttributes();
	}
	
	/*
	 * set arff loader
	 */
	public void setArffLoader() throws IOException {
		System.out.println("load arff data");
		arffLoader = new ArffLoader(); 
		arffLoader.setFile(new File(arffFile)); 
		structure = arffLoader.getStructure();
	}
	
	/*
	 * set csv saver
	 */
	public void setCsvSaver() throws IOException {
		csvSaver = new CSVSaver();
		csvSaver.setRetrieval(Saver.INCREMENTAL);
		csvSaver.setStructure(structure);
	}
	
	/*
	 * read xml file and set labels
	 */
	public void setLabels() throws DocumentException {
		SAXReader reader = new SAXReader();
		Document document = reader.read(new File(xmlFile));  
		Element root = document.getRootElement(); 
		List<Element> list = root.elements();
		for(Element element:list){
			String attrName = element.attributeValue("name");
			labels.add(attrName);
		}
		
		System.out.println("set labels " + labels.size());
	}
	
	/*
	 * set attributes
	 */
	public void setAttributes() throws IOException {
		for (int i=0; i < structure.numAttributes()-labels.size(); i++) {
			attributes.add(structure.attribute(i).name());
		}
		
		System.out.println("set attributes " + attributes.size());
	}
	
	/*
	 * save data as csv file
	 */
	public void saveAsCsv(String csvFile) throws IOException {
		System.out.println("save to csv file...");
		
		File file = new File(csvFile);
		if (file.exists()) {
			file.delete();
		}
		csvSaver.setFile(file);

		Instance instance;
		while((instance = arffLoader.getNextInstance(structure)) != null) {  
	        csvSaver.writeIncremental(instance);  
	    }
	}
	
	/*
	 * save data as txt file and label file
	 */
	public void saveAsTxtAndLabel(String dir) throws IOException {
		System.out.print("save to txt and label file...");
		
		File dirFile = new File(dir);
		if(!dirFile.exists()){
			dirFile.mkdir();
		}
		
		int count = 1;
		int numAttribute = attributes.size();
		StringBuilder builder = null;
		
		Instance instance;
		while((instance = arffLoader.getNextInstance(structure)) != null) {		
			
			if(instance.numAttributes() != (attributes.size() + labels.size())){
				continue;
			}
			
			// save the context of file
			builder = new StringBuilder("");
			int i = 0;
			while(i < numAttribute) {
				if((int)instance.value(i) == 1) {
					builder.append(attributes.get(i)).append(" ");
				}
				i++;
			}
			FileWriter writer = new FileWriter(new File(dir + "/" + count + ".txt"));
			writer.write(builder.append("\n").toString());
			writer.close();
			
			// save the labels of this file
			builder = new StringBuilder("");
			while(i < instance.numAttributes()) {
				if ((int)instance.value(i) == 1) {
					builder.append(labels.get(i-numAttribute)).append("\n");
				}
				i++;
			}
			writer = new FileWriter(new File(dir + "/" + count + ".lab"));
			writer.write(builder.toString());
			writer.close();
			
			count++;
			
		}
		
		System.out.println(count*2 + " instances");
	}

	/*
	 * save as text file
	 */
	public void saveAsText(String dir) throws IOException {
		System.out.println("save to text into labelDirs...");
		
		// create the label dir
		for (String label : labels) {
			new File(dir + "/" + label).mkdir();
		}
		
		Instance instance = null;
		while((instance = arffLoader.getNextInstance(structure)) != null) {
			if(instance.numAttributes() != (attributes.size() + labels.size())){
				continue;
			}
			
			StringBuilder builder = new StringBuilder("");
			int i = 0;
			while(i < attributes.size()) {
				if((int)instance.value(i) == 1) {
					builder.append(attributes.get(i)).append(" ");
				}
				i++;
			}
			String text = builder.append("\n").toString();
			
			FileWriter writer = null;
			while(i < instance.numAttributes()) {
				if ((int)instance.value(i) == 1) {
					int count = new File(dir + "/" + instance.attribute(i).name()).listFiles().length;
					writer = new FileWriter(new File(dir + "/" + instance.attribute(i).name() + "/" + count + ".txt"));
					writer.write(text);
					writer.close();
				}
				i++;
			}
			
		}
		
	}
	
}

**Comment by bjlgcxc** [comment]
the function saveAsTxtAndLabel() above  is used

**Comment by NowayIndustries** [comment]
Very nice! 

The question remains though, how come your testing is giving you such low values. Perhaps more/less epochs might do the trick, it all depends on what magpie is telling you I guess. Maybe you are overfitting your model, or severly underfitting it.

 Please let me know what worked/didn't work to get the prediction/testing values up so I can maybe save myself some trouble in a couple of weeks ;).

**Comment by bjlgcxc** [comment]
ok,I'll let you know if I solve the problem

**Comment by jstypka** [comment]
@bjlgcxc is the validation loss (the error on the test set) decreasing at all between epochs? Could you post how your `.txt` and `.lab` files look like? Make sure they are structured in the same way as the example dataset [here](https://github.com/inspirehep/magpie/tree/master/data/hep-keywords/data/train).

**Comment by bjlgcxc** [comment]
@jstypka 
the txt:
     a abstract act also an and applying architecture architectures as base be building built by can capture common complex complexity computer considered contribute defining describe design development distribution experience for from how identify identifying in introduce many mechanism more n new notes object objects of oriented our overall patterns play process propose provide reduce reuse roles science software system systems that the their they to we which 

the lab:
    TAG_design
    TAG_kaldesignresearch
    TAG_patterns

**Comment by jstypka** [comment]
@bjlgcxc the format looks ok! One interesting thing is that you already filtered out the stopwords etc - this should probably give better results for other methods like bag-of-words, but I don't think it would improve results which are using `word2vec` models. You might be better off with feeding full text as an input and leaving the rest to the machine.

**Comment by bjlgcxc** [comment]
@jstypka  ok,thank you very much

**Comment by degorsse** [comment]
Hi Jan, I just read your message concerning stopwords. We also eliminated the stopwords in different trials. Can you explain why for word2vec performance may be lowered? On a general matter is there any (other) cleaning operations to do to the full text ? (ex: Single sentences better than full text ? / Eliminating @ and other special char. ? etc.)

**Comment by jstypka** [comment]
`word2vec` relies on the context of each word to figure out it's meaning, so for example if it spots the preposition _to_ before some words in might conclude that it is a place and draw a relation between them in the model. If you strip words like _to_ then you take away this information.

https://www.quora.com/Is-it-compulsory-to-remove-stop-words-with-word2vec

But perhaps removing them doesn't hurt the results - frankly, I have never tried running it without stopwords.
